plugins/generator-1.17.1/forge-1.17.1/templates/procedure.java.ftl

```diff
@@ -33,29 +33,49 @@ package ${package}.procedures;
 
 import net.minecraftforge.eventbus.api.Event;
 
+<#assign nullableDependencies = []/>
+<#list dependencies as dependency>
+	<#if dependency.getType(generator.getWorkspace()) != "double"
+		&& dependency.getType(generator.getWorkspace()) != "LevelAccessor"
+		&& dependency.getType(generator.getWorkspace()) != "ItemStack"
+		&& dependency.getType(generator.getWorkspace()) != "BlockState"
+		&& dependency.getType(generator.getWorkspace()) != "InteractionResult"
+		&& dependency.getType(generator.getWorkspace()) != "boolean">
+		<#assign nullableDependencies += [dependency.getName()]/>
+	</#if>
+</#list>
+
+<#compress>
+
 <#if trigger_code?has_content>
 ${trigger_code}
 <#else>
 public class ${name}Procedure {
 </#if>
 
-	public static <#if return_type??>${return_type.getJavaType(generator.getWorkspace())}<#else>void</#if> execute(Map<String, Object> dependencies) {
+	<#if trigger_code?has_content>
+	public static <#if return_type??>${return_type.getJavaType(generator.getWorkspace())}<#else>void</#if> execute(
 		<#list dependencies as dependency>
-			if(dependencies.get("${dependency.getName()}") == null) {
-				if(!dependencies.containsKey("${dependency.getName()}"))
-					${JavaModName}.LOGGER.warn("Failed to load dependency ${dependency.getName()} for procedure ${name}!");
-				<#if return_type??>return ${return_type.getDefaultValue(generator.getWorkspace())}<#else>return</#if>;
-			}
-        </#list>
+			${dependency.getType(generator.getWorkspace())} ${dependency.getName()}<#if dependency?has_next>,</#if>
+		</#list>
+	) {
+		<#if return_type??>return </#if>execute(null<#if dependencies?has_content>,</#if><#list dependencies as dependency>${dependency.getName()}<#if dependency?has_next>,</#if></#list>);
+	}
+	</#if>
 
+	<#if trigger_code?has_content>private <#else>public </#if>static <#if return_type??>${return_type.getJavaType(generator.getWorkspace())}<#else>void</#if> execute(
+		<#if trigger_code?has_content>@Nullable Event event<#if dependencies?has_content>,</#if></#if>
 		<#list dependencies as dependency>
-			<#if dependency.getType(generator.getWorkspace()) == "double">
-				double ${dependency.getName()} = dependencies.get("${dependency.getName()}") instanceof Integer
-					? (int) dependencies.get("${dependency.getName()}") : (double) dependencies.get("${dependency.getName()}");
-			<#else>
-				${dependency.getType(generator.getWorkspace())} ${dependency.getName()} = (${dependency.getType(generator.getWorkspace())}) dependencies.get("${dependency.getName()}");
-			</#if>
+				${dependency.getType(generator.getWorkspace())} ${dependency.getName()}<#if dependency?has_next>,</#if>
 		</#list>
+	) {
+		<#if nullableDependencies?has_content>
+			if(
+			<#list nullableDependencies as dependency>
+			${dependency} == null <#if dependency?has_next>||</#if>
+			</#list>
+			) return <#if return_type??>${return_type.getDefaultValue(generator.getWorkspace())}</#if>;

```

Comment:
Why not just reassign the nulled dependency to its type default value?

Reply:
What is default value for eg. Entity object?

Reply:
I'd say you can still use mod logger in such cases.

Reply:
Before it was logged if a dependency was missing in the map, not if it was null only. Now if dependency will be missing, code simply won't compile which is much more straight forward and reduces confusion for the end-user

Reply:
Then the final question: would you make it reassign primitive type dependencies?

Reply:
OKay, nevermind.

Reply:
It does not reassign them I think